<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Introduction: Bioconductor S4 Objects / SummarizedExperiment</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="01_introduction_files/libs/clipboard/clipboard.min.js"></script>
<script src="01_introduction_files/libs/quarto-html/quarto.js"></script>
<script src="01_introduction_files/libs/quarto-html/popper.min.js"></script>
<script src="01_introduction_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="01_introduction_files/libs/quarto-html/anchor.min.js"></script>
<link href="01_introduction_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01_introduction_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="01_introduction_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="01_introduction_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="01_introduction_files/libs/bootstrap/bootstrap-0b9e35696ea153a55a49e2492062f191.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><link href="01_introduction_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="01_introduction_files/libs/pagedtable-1.1/js/pagedtable.js"></script><link rel="stylesheet" href="style.css">
</head>
<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#general-purpose-of-rnaseq-experiments" id="toc-general-purpose-of-rnaseq-experiments" class="nav-link" data-scroll-target="#general-purpose-of-rnaseq-experiments">General Purpose of RNAseq Experiments</a></li>
  <li><a href="#what-are-we-trying-to-accomplish" id="toc-what-are-we-trying-to-accomplish" class="nav-link" data-scroll-target="#what-are-we-trying-to-accomplish">What are we trying to accomplish?</a></li>
  <li><a href="#experimental-design-of-rnaseq-experiments" id="toc-experimental-design-of-rnaseq-experiments" class="nav-link" data-scroll-target="#experimental-design-of-rnaseq-experiments">Experimental Design of RNAseq Experiments</a></li>
  <li><a href="#what-is-bioconductor" id="toc-what-is-bioconductor" class="nav-link" data-scroll-target="#what-is-bioconductor">What is Bioconductor?</a></li>
  <li>
<a href="#s4-bioconductor-objects" id="toc-s4-bioconductor-objects" class="nav-link" data-scroll-target="#s4-bioconductor-objects">S4 Bioconductor Objects</a>
  <ul class="collapse">
<li><a href="#validation" id="toc-validation" class="nav-link" data-scroll-target="#validation">Validation</a></li>
  </ul>
</li>
  <li><a href="#summarizedexperiment" id="toc-summarizedexperiment" class="nav-link" data-scroll-target="#summarizedexperiment"><code>SummarizedExperiment</code></a></li>
  <li><a href="#assay-data" id="toc-assay-data" class="nav-link" data-scroll-target="#assay-data">Assay Data</a></li>
  <li><a href="#metadata" id="toc-metadata" class="nav-link" data-scroll-target="#metadata">Metadata</a></li>
  <li><a href="#subsetting-summarizedexperiment" id="toc-subsetting-summarizedexperiment" class="nav-link" data-scroll-target="#subsetting-summarizedexperiment">Subsetting <code>SummarizedExperiment</code></a></li>
  <li>
<a href="#tidysummarizedexperiment" id="toc-tidysummarizedexperiment" class="nav-link" data-scroll-target="#tidysummarizedexperiment"><code>tidySummarizedExperiment</code></a>
  <ul class="collapse">
<li><a href="#filtering-data" id="toc-filtering-data" class="nav-link" data-scroll-target="#filtering-data">Filtering Data</a></li>
  <li><a href="#getting-counts-across-the-experiment" id="toc-getting-counts-across-the-experiment" class="nav-link" data-scroll-target="#getting-counts-across-the-experiment">Getting Counts across the experiment</a></li>
  <li><a href="#try-it-out" id="toc-try-it-out" class="nav-link" data-scroll-target="#try-it-out">Try it out</a></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting">Plotting</a></li>
  </ul>
</li>
  <li><a href="#take-home-messages" id="toc-take-home-messages" class="nav-link" data-scroll-target="#take-home-messages">Take Home Messages</a></li>
  </ul></nav>
</div>
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Introduction: Bioconductor S4 Objects / <code>SummarizedExperiment</code>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="learning-objectives" class="level2"><h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>
<strong>Explain</strong> the basics of how RNAseq experiments are conducted</li>
<li>
<strong>Explain</strong> the basics of experimental design, batch effects, and when to consult the sequencing core</li>
<li>
<strong>Distinguish</strong> between <em>technical</em> and <em>biological</em> replicates and what kinds of results to expect from each</li>
<li>
<strong>Explain</strong> the basic structure of Bioconductor S4 objects</li>
<li>
<strong>Explain</strong> and <strong>Utilize</strong> <code>SummarizedExperiment</code> objects</li>
</ul></section><section id="general-purpose-of-rnaseq-experiments" class="level2"><h2 class="anchored" data-anchor-id="general-purpose-of-rnaseq-experiments">General Purpose of RNAseq Experiments</h2>
<p>Bulk RNAseq experiments are meant to distinguish expression differences between samples or groups of samples.</p>
<p><a href="00_concepts.html">Please see the concepts section</a> for more in depth details of the processing of RNA-seq experiments.</p>
</section><section id="what-are-we-trying-to-accomplish" class="level2"><h2 class="anchored" data-anchor-id="what-are-we-trying-to-accomplish">What are we trying to accomplish?</h2>
<p>In RNAseq, we are interested in assessing <em>expression differences</em> between samples at either the gene or transcript level.</p>
<p>In order to do this, we need <em>phenotype information</em> about the samples in our experiment. Specifically, we need to have information about:</p>
<ul>
<li>conditions (Tissue type)</li>
<li>possible confounders (gender, time)</li>
</ul></section><section id="experimental-design-of-rnaseq-experiments" class="level2"><h2 class="anchored" data-anchor-id="experimental-design-of-rnaseq-experiments">Experimental Design of RNAseq Experiments</h2>
<p>When we design our RNAseq experiment, we want to be able to:</p>
<ul>
<li>Number and type of <strong>replicates</strong>
</li>
<li>Avoiding <strong>confounding</strong>
</li>
<li>Addressing <strong>batch effects</strong>
</li>
</ul>
<p>The important thing to note is that if you’re unsure about the experimental design, <em>contact the Bioinformatics Core</em>. They can help you design the experiment based on what samples you have.</p>
<p>We’ll talk much more about experimental design next week.</p>
</section><section id="what-is-bioconductor" class="level2"><h2 class="anchored" data-anchor-id="what-is-bioconductor">What is Bioconductor?</h2>
<p>Data in bioinformatics is often complex. To deal with this, developers define specialised data containers (termed classes) that match the properties of the data they need to handle.</p>
<p>This aspect is central to the <strong>Bioconductor</strong><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> project which uses the same <strong>core data infrastructure</strong> across packages. This certainly contributed to Bioconductor’s success. Bioconductor package developers are advised to make use of existing infrastructure to provide coherence, interoperability, and stability to the project as a whole.</p>
</section><section id="s4-bioconductor-objects" class="level2"><h2 class="anchored" data-anchor-id="s4-bioconductor-objects">S4 Bioconductor Objects</h2>
<p>At the heart of Bioconductor are the <a href="https://bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html">S4 Objects</a>. You can think of these data structures as ways to package together:</p>
<ul>
<li>Assay Data, such as an RNAseq count matrix</li>
<li>Metadata (data about the experiment), including Experimental Design</li>
</ul>
<p>Part of why Bioconductor works is this data packaging. We can write routines to act on the data in these S4 Objects as part of a processing <em>workflow</em>. As long as we output a Bioconductor S4 object, our routines can work as part of a pipeline. These routines are called <code><a href="https://rdrr.io/r/utils/methods.html">methods()</a></code>, and may come from a variety of packages.</p>
<p>You can think of the Bioconductor S4 objects as taking the place of <code>data.frame</code>s in <code>dplyr</code> pipelines - they are the common format that all of the Bioconductor methods work on. They allow Bioconductor methods to be <code>interoperable</code>.</p>
<section id="validation" class="level3"><h3 class="anchored" data-anchor-id="validation">Validation</h3>
<p>The other big part of S4 objects is <em>validation</em>. This can be a bit hard to wrap your head around.</p>
<p>The Bioconductor designers put special validation checks on the input data for the Bioconductor objects when you load data into them. The following is what is called the <em>Constructor</em> for the <code>SummarizedExperiment</code> object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">SummarizedExperiment</span><span class="op">(</span>assays<span class="op">=</span><span class="fu">SimpleList</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     rowData<span class="op">=</span><span class="cn">NULL</span>, rowRanges<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                     colData<span class="op">=</span><span class="fu">DataFrame</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     metadata<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     checkDimnames<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Each argument to the <code>SummarizedExperiment</code> constructor defines restrictions on that slot.</p>
<p>The <code>checkDimnames</code> argument is critical. In order to do any work with an experiment you need to map samples to <code>colNames</code> (the experimental matrix). For example, to calculate differential expression between samples, you need to specify the different groups to compare and which samples map to which groups. Thus, the column names in the AssayData must be identical to the row names in <code>colData</code>.</p>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
If it isn’t in <code>colData</code>, it doesn’t exist
</div>
</div>
<div class="callout-body-container callout-body">
<p>Keep in mind that your <code>colData</code> must be as complete as possible. Why? The short answer is that if there isn’t a row for your sample in <code>colData</code>, then it basically doesn’t exist for Bioconductor methods.</p>
<p>So make sure your <code>colData</code> contains sample names for all your samples.</p>
</div>
</div>
</section></section><section id="summarizedexperiment" class="level2"><h2 class="anchored" data-anchor-id="summarizedexperiment"><code>SummarizedExperiment</code></h2>
<p>The following diagram shows how the different <em>data slots</em> in the <code>SummarizedExperiment</code> relate to each other.</p>
<p><img src="images/SE.svg" class="img-fluid"></p>
<p>We’ll take a look at data in a <code>SummarizedExperiment</code> object:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSE96870</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">GSE96870</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 41786 22 
metadata(0):
assays(1): counts
rownames(41786): Xkr4 LOC105243853 ... TrnT TrnP
rowData names(3): ENTREZID product gbkey
colnames(22): GSM2545337 GSM2545338 ... GSM2545346 GSM2545347
colData names(12): title geo_accession ... Label Group</code></pre>
</div>
</div>
<p>In some ways, the <code>SummarizedExperiment</code> object is like a <code>data.frame</code>, but with extra metadata. For example, our object has column names, which correspond to sample identifiers:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">GSE96870</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "GSM2545337" "GSM2545338" "GSM2545348" "GSM2545353" "GSM2545343"
 [6] "GSM2545349" "GSM2545354" "GSM2545339" "GSM2545344" "GSM2545352"
[11] "GSM2545362" "GSM2545340" "GSM2545345" "GSM2545350" "GSM2545363"
[16] "GSM2545336" "GSM2545342" "GSM2545351" "GSM2545380" "GSM2545341"
[21] "GSM2545346" "GSM2545347"</code></pre>
</div>
</div>
<p>And row names:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">GSE96870</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Xkr4"          "LOC105243853"  "LOC105242387"  "LOC105242467" 
 [5] "Rp1"           "Sox17"         "Gm7357"        "LOC105243855" 
 [9] "LOC105243854"  "Gm7369"        "Gm29874"       "Gm6123"       
[13] "Mrpl15"        "Lypla1"        "LOC105243856"  "Tcea1"        
[17] "Gm6104"        "Rgs20"         "Gm16041"       "Atp6v1h"      
[21] "Oprk1"         "Npbwr1"        "4732440D04Rik" "Rb1cc1"       
[25] "Fam150a"       "Gm2147"        "LOC102631893"  "St18"         
[29] "LOC102641523"  "Pcmtd1"       </code></pre>
</div>
</div>
<p>So far, so good.</p>
</section><section id="assay-data" class="level2"><h2 class="anchored" data-anchor-id="assay-data">Assay Data</h2>
<p>Everything in <code>SummarizedExperiment</code> is built around the Assay data that we store in it.</p>
<p>Each element of the Assay data list contains a <code>data.frame</code> with the following contents:</p>
<ul>
<li>
<strong>Rows</strong>: correspond to Gene Locus or possibly Transcripts</li>
<li>
<strong>Columns</strong>: correspond to the samples used in your experiment.</li>
<li>
<strong>Values</strong>: correspond to the actual data. In the case of RNAseq, these are the <em>counts</em> that map to each row.</li>
</ul>
<p>You can extract the assay data using the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay()</a></code> method:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">assay_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">GSE96870</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">assay_data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             GSM2545337 GSM2545338 GSM2545348 GSM2545353 GSM2545343 GSM2545349
Xkr4               2410       2159       2275       1910       2235       1881
LOC105243853          0          1          1          0          3          0
LOC105242387        121        110        161        214        130        154
LOC105242467          5          5          2          1          2          4
Rp1                   2          0          3          1          1          6
Sox17               239        218        302        322        296        286
             GSM2545354 GSM2545339 GSM2545344 GSM2545352 GSM2545362 GSM2545340
Xkr4               1771       1980       1779       1890       2315       1977
LOC105243853          0          4          3          1          1          0
LOC105242387        124        120        131        272        189        172
LOC105242467          4          5          2          3          2          2
Rp1                   3          3          1          5          3          2
Sox17               273        220        233        267        197        261
             GSM2545345 GSM2545350 GSM2545363 GSM2545336 GSM2545342 GSM2545351
Xkr4               1528       2584       1645       1891       1757       1837
LOC105243853          0          0          0          0          1          1
LOC105242387        160        124        223        204        177        221
LOC105242467          2          7          1         12          3          1
Rp1                   2          5          1          2          3          3
Sox17               271        325        310        251        179        201
             GSM2545380 GSM2545341 GSM2545346 GSM2545347
Xkr4               1723       1945       1644       1585
LOC105243853          1          0          1          3
LOC105242387        251        173        180        176
LOC105242467          4          6          1          2
Rp1                   0          1          2          2
Sox17               246        232        205        230</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Assays Are Not just for RNAseq data
</div>
</div>
<div class="callout-body-container callout-body">
<p>Assay Data is very flexible. For example, there are flow cytometry objects where the rows correspond to cell surface markers, and columns that correspond to each cell. Similarly, <code>SingleCellExperiment</code> objects (which are derived from <code>SummarizedExperiment</code>) have rows that correspond to Genes and columns that correspond to individual cells.</p>
</div>
</div>
</section><section id="metadata" class="level2"><h2 class="anchored" data-anchor-id="metadata">Metadata</h2>
<p>Metadata is information about the experiment that is not part of the Assay Data.</p>
<p>The most important part of the metadata is the <code>colData</code> slot. This slot contains information about the samples (the columns) of the assay. This is where we store the Experimental Design that we talked about.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">GSE96870</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 22 rows and 12 columns
                     title geo_accession     organism         age         sex
               &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt; &lt;character&gt;
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus     8 weeks      Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus     8 weeks      Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus     8 weeks      Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus     8 weeks      Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus     8 weeks        Male
...                    ...           ...          ...         ...         ...
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus     8 weeks      Female
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus     8 weeks      Female
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus     8 weeks        Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus     8 weeks        Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus     8 weeks        Male
             infection      strain        time      tissue     mouse
           &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt;
GSM2545337 NonInfected     C57BL/6        Day0  Cerebellum         9
GSM2545338 NonInfected     C57BL/6        Day0  Cerebellum        10
GSM2545348 NonInfected     C57BL/6        Day0  Cerebellum         8
GSM2545353 NonInfected     C57BL/6        Day0  Cerebellum         4
GSM2545343 NonInfected     C57BL/6        Day0  Cerebellum        11
...                ...         ...         ...         ...       ...
GSM2545351  InfluenzaA     C57BL/6        Day8  Cerebellum        16
GSM2545380  InfluenzaA     C57BL/6        Day8  Cerebellum        19
GSM2545341  InfluenzaA     C57BL/6        Day8  Cerebellum         6
GSM2545346  InfluenzaA     C57BL/6        Day8  Cerebellum        23
GSM2545347  InfluenzaA     C57BL/6        Day8  Cerebellum        24
                    Label       Group
                 &lt;factor&gt;    &lt;factor&gt;
GSM2545337 Female_Day0_9  Female_Day0
GSM2545338 Female_Day0_10 Female_Day0
GSM2545348 Female_Day0_8  Female_Day0
GSM2545353 Female_Day0_4  Female_Day0
GSM2545343 Male_Day0_11   Male_Day0  
...                   ...         ...
GSM2545351 Female_Day8_16 Female_Day8
GSM2545380 Female_Day8_19 Female_Day8
GSM2545341 Male_Day8_6    Male_Day8  
GSM2545346 Male_Day8_23   Male_Day8  
GSM2545347 Male_Day8_24   Male_Day8  </code></pre>
</div>
</div>
<p>In our experimental design, we have males and females, timepoints, and different kinds of tissues.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Keep your <code>SummarizedExperiment</code> whole
</div>
</div>
<div class="callout-body-container callout-body">
<p>It might be tempting to extract the assay data and the metadata and work with them separately. But as we’ll see in the following section, these two slots work together and enable all sorts of analysis.</p>
</div>
</div>
</section><section id="subsetting-summarizedexperiment" class="level2"><h2 class="anchored" data-anchor-id="subsetting-summarizedexperiment">Subsetting <code>SummarizedExperiment</code>
</h2>
<p>We saw that we have the <code>checkNames</code> constraint. This is because we can use the <code>colData</code> and the <code>assayData</code> in our object to do subsetting using the metadata.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">female</span> <span class="op">&lt;-</span> <span class="va">GSE96870</span><span class="op">[</span>,<span class="va">GSE96870</span><span class="op">$</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span><span class="op">]</span></span>
<span><span class="va">female</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 41786 12 
metadata(0):
assays(1): counts
rownames(41786): Xkr4 LOC105243853 ... TrnT TrnP
rowData names(3): ENTREZID product gbkey
colnames(12): GSM2545337 GSM2545338 ... GSM2545351 GSM2545380
colData names(12): title geo_accession ... Label Group</code></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">female</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 12 rows and 12 columns
                     title geo_accession     organism         age         sex
               &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt; &lt;character&gt;
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus     8 weeks      Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus     8 weeks      Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus     8 weeks      Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus     8 weeks      Female
GSM2545339 CNS_RNA-seq_13C    GSM2545339 Mus musculus     8 weeks      Female
...                    ...           ...          ...         ...         ...
GSM2545362  CNS_RNA-seq_5C    GSM2545362 Mus musculus     8 weeks      Female
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus     8 weeks      Female
GSM2545342  CNS_RNA-seq_1C    GSM2545342 Mus musculus     8 weeks      Female
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus     8 weeks      Female
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus     8 weeks      Female
             infection      strain        time      tissue     mouse
           &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt;
GSM2545337 NonInfected     C57BL/6        Day0  Cerebellum         9
GSM2545338 NonInfected     C57BL/6        Day0  Cerebellum        10
GSM2545348 NonInfected     C57BL/6        Day0  Cerebellum         8
GSM2545353 NonInfected     C57BL/6        Day0  Cerebellum         4
GSM2545339  InfluenzaA     C57BL/6        Day4  Cerebellum        15
...                ...         ...         ...         ...       ...
GSM2545362  InfluenzaA     C57BL/6        Day4  Cerebellum        20
GSM2545336  InfluenzaA     C57BL/6        Day8  Cerebellum        14
GSM2545342  InfluenzaA     C57BL/6        Day8  Cerebellum         5
GSM2545351  InfluenzaA     C57BL/6        Day8  Cerebellum        16
GSM2545380  InfluenzaA     C57BL/6        Day8  Cerebellum        19
                    Label       Group
                 &lt;factor&gt;    &lt;factor&gt;
GSM2545337 Female_Day0_9  Female_Day0
GSM2545338 Female_Day0_10 Female_Day0
GSM2545348 Female_Day0_8  Female_Day0
GSM2545353 Female_Day0_4  Female_Day0
GSM2545339 Female_Day4_15 Female_Day4
...                   ...         ...
GSM2545362 Female_Day4_20 Female_Day4
GSM2545336 Female_Day8_14 Female_Day8
GSM2545342 Female_Day8_5  Female_Day8
GSM2545351 Female_Day8_16 Female_Day8
GSM2545380 Female_Day8_19 Female_Day8</code></pre>
</div>
</div>
<p>This tight interaction of metadata and assay data is critical when we start doing differential analysis. The experimental design will help determine whether we can make the comparisons we want to make and the conclusions we can draw from the dataset.</p>
</section><section id="tidysummarizedexperiment" class="level2"><h2 class="anchored" data-anchor-id="tidysummarizedexperiment"><code>tidySummarizedExperiment</code></h2>
<p>The <code>SummarizedExperiment</code> object does not act like the normal <code>data.frame</code>/<code>tibble</code> we expect, especially in filtering and subsetting.</p>
<p>The <code>tidySummarizedExperiment</code> package will be helpful for us to understand and visualize the <code>SummarizedExperiment</code> package. Once we do that, our <code>SummarizedExperiment</code> object will act more like a <code>data.frame</code>.</p>
<p>Basically, <code>tidySummarizedExperiment</code> treats the data as a long data frame that combines both the assay and metadata. This format is helpful in doing more work with the <code>tidyverse</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySummarizedExperiment">tidySummarizedExperiment</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ttservice</code></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 919,292 × 23
# Features=41786 | Samples=22 | Assays=counts
   .feature    .sample counts title geo_accession organism age   sex   infection
   &lt;chr&gt;       &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    
 1 Xkr4        GSM254…   2410 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 2 LOC1052438… GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 3 LOC1052423… GSM254…    121 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 4 LOC1052424… GSM254…      5 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 5 Rp1         GSM254…      2 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 6 Sox17       GSM254…    239 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 7 Gm7357      GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 8 LOC1052438… GSM254…      3 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 9 LOC1052438… GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
10 Gm7369      GSM254…      1 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
# ℹ 40 more rows
# ℹ 14 more variables: strain &lt;chr&gt;, time &lt;chr&gt;, tissue &lt;chr&gt;, mouse &lt;int&gt;,
#   Label &lt;fct&gt;, Group &lt;fct&gt;, ENTREZID &lt;chr&gt;, product &lt;chr&gt;, gbkey &lt;chr&gt;,
#   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;</code></pre>
</div>
</div>
<section id="filtering-data" class="level3"><h3 class="anchored" data-anchor-id="filtering-data">Filtering Data</h3>
<p>We can use <code>dplyr</code> to filter the data using the <code>tidySummarizedExperiment</code> package:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 501,432 × 23
# Features=41786 | Samples=12 | Assays=counts
   .feature    .sample counts title geo_accession organism age   sex   infection
   &lt;chr&gt;       &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    
 1 Xkr4        GSM254…   2410 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 2 LOC1052438… GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 3 LOC1052423… GSM254…    121 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 4 LOC1052424… GSM254…      5 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 5 Rp1         GSM254…      2 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 6 Sox17       GSM254…    239 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 7 Gm7357      GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 8 LOC1052438… GSM254…      3 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 9 LOC1052438… GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
10 Gm7369      GSM254…      1 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
# ℹ 40 more rows
# ℹ 14 more variables: strain &lt;chr&gt;, time &lt;chr&gt;, tissue &lt;chr&gt;, mouse &lt;int&gt;,
#   Label &lt;fct&gt;, Group &lt;fct&gt;, ENTREZID &lt;chr&gt;, product &lt;chr&gt;, gbkey &lt;chr&gt;,
#   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;</code></pre>
</div>
</div>
</section><section id="getting-counts-across-the-experiment" class="level3"><h3 class="anchored" data-anchor-id="getting-counts-across-the-experiment">Getting Counts across the experiment</h3>
<p>We can produce summaries of the counts of each sample library:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">.sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">summarise</span><span class="op">(</span>total_counts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>tidySummarizedExperiment says: A data frame is returned for independent data analysis.</code></pre>
</div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[".sample"],"name":[1],"type":["chr"],"align":["left"]},{"label":["total_counts"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"GSM2545336","2":"40265225"},{"1":"GSM2545337","2":"33818586"},{"1":"GSM2545338","2":"31684299"},{"1":"GSM2545339","2":"32663438"},{"1":"GSM2545340","2":"33518799"},{"1":"GSM2545341","2":"32777049"},{"1":"GSM2545342","2":"31624027"},{"1":"GSM2545343","2":"40031023"},{"1":"GSM2545344","2":"33726110"},{"1":"GSM2545345","2":"34043596"},{"1":"GSM2545346","2":"33642930"},{"1":"GSM2545347","2":"36130719"},{"1":"GSM2545348","2":"37698775"},{"1":"GSM2545349","2":"36643928"},{"1":"GSM2545350","2":"43409628"},{"1":"GSM2545351","2":"37224356"},{"1":"GSM2545352","2":"43189078"},{"1":"GSM2545353","2":"39520008"},{"1":"GSM2545354","2":"35470183"},{"1":"GSM2545362","2":"38546511"},{"1":"GSM2545363","2":"36499016"},{"1":"GSM2545380","2":"40921490"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section><section id="try-it-out" class="level3"><h3 class="anchored" data-anchor-id="try-it-out">Try it out</h3>
<p>Try summarizing the <code>total_counts</code> by <code>sex</code> or by <code>infection</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 919,292 × 23
# Features=41786 | Samples=22 | Assays=counts
   .feature    .sample counts title geo_accession organism age   sex   infection
   &lt;chr&gt;       &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    
 1 Xkr4        GSM254…   2410 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 2 LOC1052438… GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 3 LOC1052423… GSM254…    121 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 4 LOC1052424… GSM254…      5 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 5 Rp1         GSM254…      2 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 6 Sox17       GSM254…    239 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 7 Gm7357      GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 8 LOC1052438… GSM254…      3 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
 9 LOC1052438… GSM254…      0 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
10 Gm7369      GSM254…      1 CNS_… GSM2545337    Mus mus… 8 we… Fema… NonInfec…
# ℹ 40 more rows
# ℹ 14 more variables: strain &lt;chr&gt;, time &lt;chr&gt;, tissue &lt;chr&gt;, mouse &lt;int&gt;,
#   Label &lt;fct&gt;, Group &lt;fct&gt;, ENTREZID &lt;chr&gt;, product &lt;chr&gt;, gbkey &lt;chr&gt;,
#   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;</code></pre>
</div>
</div>
</section><section id="plotting" class="level3"><h3 class="anchored" data-anchor-id="plotting">Plotting</h3>
<p>We can look at the distribution of counts across samples by using <code>.sample</code> as our grouping variable in <code>ggplot()</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group<span class="op">=</span><span class="va">.sample</span>, color<span class="op">=</span><span class="va">infection</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="01_introduction_files/figure-html/tidySE-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Our data (individual gene counts) is highly skewed. If we log-transform the data, then the skew is lessened. The boxplots of log counts are helpful for detecting batch effects.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, group<span class="op">=</span><span class="va">.sample</span>, fill<span class="op">=</span><span class="va">infection</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="01_introduction_files/figure-html/tidy-boxplot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Take a look at the above box plot. Are there differences in distribution between samples?</p>
<p>In order to compare expression levels between samples, we will need to adjust, or <em>normalize</em> by library size. We’ll investigate this further when we get to differential expression.</p>
<p>Try mapping another variable to <code>fill</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>GSE96870 <span class="sc">|&gt;</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span><span class="fu">log2</span>(counts <span class="sc">+</span> <span class="dv">1</span>), <span class="at">group=</span>.sample, <span class="at">fill=</span><span class="sc">------</span>)) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="take-home-messages" class="level2"><h2 class="anchored" data-anchor-id="take-home-messages">Take Home Messages</h2>
<ul>
<li>We package multiple parts of the data (assay data and metadata) into a <code>SummarizedExperiment</code> object:
<ul>
<li>
<strong>Columns</strong> correspond to <em>samples</em>, <strong>Rows</strong> correspond to <em>gene locus</em> or <em>transcripts</em>
</li>
<li>
<em>Column names</em> of the assay data must match a column in the <code>colData</code>
</li>
<li>
<em>Experimental Design</em> variables are part of the <code>colData</code>
</li>
<li>Rownames of the assay data must match the rowNames in the column data</li>
</ul>
</li>
<li>The <code>SummarizedExperiment</code> object lets us subset by phenotype variables</li>
<li>
<code>tidySummarizedExperiment</code> lets us interact with the <code>SummarizedExperiment</code> object as if it were a long data frame with both metadata and assay data together.
<ul>
<li>We can plot and subset the <code>SummarizedExperiment</code> object using our regular <code>tidyverse</code> tools (<code>ggplot2</code>, <code>dplyr</code>, etc.) once we load the <code>tidySummarizedExperiment</code> package.</li>
</ul>
</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What about all of the other kinds of objects?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most of the other Bioconductor Data Structures derive from some variant of the <code>SummarizedExperiment</code>. They might add some functionality that is core to the package they belong to, such as <code>DEseq2</code>. This includes <code>seurat</code> objects for Single Cell sequencing.</p>
</div>
</div>
</section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p>The <a href="https://www.bioconductor.org">Bioconductor</a> project was initiated by Robert Gentleman, one of the two creators of the R language. Bioconductor provides tools dedicated to omics data analysis. Bioconductor uses the R statistical programming language and is open source and open development.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></main><!-- /main column --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>