<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Experimental Design / QC</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="02_week2_files/libs/clipboard/clipboard.min.js"></script>
<script src="02_week2_files/libs/quarto-html/quarto.js"></script>
<script src="02_week2_files/libs/quarto-html/popper.min.js"></script>
<script src="02_week2_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="02_week2_files/libs/quarto-html/anchor.min.js"></script>
<link href="02_week2_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="02_week2_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="02_week2_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="02_week2_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="02_week2_files/libs/bootstrap/bootstrap-0b9e35696ea153a55a49e2492062f191.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><link rel="stylesheet" href="style.css">
</head>
<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#experimental-design-of-rnaseq-experiments" id="toc-experimental-design-of-rnaseq-experiments" class="nav-link" data-scroll-target="#experimental-design-of-rnaseq-experiments">Experimental Design of RNAseq Experiments</a></li>
  <li>
<a href="#background-what-are-we-estimating" id="toc-background-what-are-we-estimating" class="nav-link" data-scroll-target="#background-what-are-we-estimating">Background: What are we estimating?</a>
  <ul class="collapse">
<li><a href="#single-randomized-group" id="toc-single-randomized-group" class="nav-link" data-scroll-target="#single-randomized-group">Single, randomized group</a></li>
  <li><a href="#two-group-comparisons" id="toc-two-group-comparisons" class="nav-link" data-scroll-target="#two-group-comparisons">Two-group comparisons</a></li>
  </ul>
</li>
  <li><a href="#dealing-with-confounding" id="toc-dealing-with-confounding" class="nav-link" data-scroll-target="#dealing-with-confounding">Dealing with confounding</a></li>
  <li>
<a href="#specific-confounding-issues-in-rna-seq-experiments" id="toc-specific-confounding-issues-in-rna-seq-experiments" class="nav-link" data-scroll-target="#specific-confounding-issues-in-rna-seq-experiments">Specific confounding issues in RNA-seq experiments</a>
  <ul class="collapse">
<li><a href="#library-size" id="toc-library-size" class="nav-link" data-scroll-target="#library-size">Library Size</a></li>
  <li><a href="#rna-composition" id="toc-rna-composition" class="nav-link" data-scroll-target="#rna-composition">RNA Composition</a></li>
  </ul>
</li>
  <li><a href="#putting-this-together-in-practice" id="toc-putting-this-together-in-practice" class="nav-link" data-scroll-target="#putting-this-together-in-practice">Putting this together in practice</a></li>
  <li><a href="#appendix-how-do-i-know-how-many-samples-to-use" id="toc-appendix-how-do-i-know-how-many-samples-to-use" class="nav-link" data-scroll-target="#appendix-how-do-i-know-how-many-samples-to-use">Appendix: How do I know how many samples to use?</a></li>
  <li><a href="#appendix-dimension-reduction-to-explore-your-data" id="toc-appendix-dimension-reduction-to-explore-your-data" class="nav-link" data-scroll-target="#appendix-dimension-reduction-to-explore-your-data">Appendix: Dimension Reduction to explore your data</a></li>
  <li><a href="#appendix-other-confounding-factors-to-consider" id="toc-appendix-other-confounding-factors-to-consider" class="nav-link" data-scroll-target="#appendix-other-confounding-factors-to-consider">Appendix: Other confounding factors to consider</a></li>
  </ul></nav>
</div>
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Experimental Design / QC</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="learning-objectives" class="level2"><h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>
<strong>Explain</strong> the the relationship between population mean vs.&nbsp;sample mean as the sample size increase</li>
<li>
<strong>Distinguish</strong> between <em>technical</em> and <em>biological</em> replicates and what kinds of results to expect from each</li>
<li>
<strong>Explain</strong> the basics of experimental design, confounding, batch effects, and when to consult the sequencing core</li>
<li>
<strong>Explain</strong> the confounding issues of library size and RNA composition</li>
<li>
<strong>Construct</strong> a design matrix that reflects the study’s experimental design</li>
<li>
<strong>Analyze</strong> the process of filtering out genes with low read counts</li>
</ul></section><section id="experimental-design-of-rnaseq-experiments" class="level2"><h2 class="anchored" data-anchor-id="experimental-design-of-rnaseq-experiments">Experimental Design of RNAseq Experiments</h2>
<p>Today, we will look how experimental design plays a large role in the way we analyze RNA-seq data. Experimental design seeks to make sense data variation via careful design of conditions.</p>
<p>If you are not familiar with some of these technical terms in the quote, we will go over them shortly.</p>
</section><section id="background-what-are-we-estimating" class="level2"><h2 class="anchored" data-anchor-id="background-what-are-we-estimating">Background: What are we estimating?</h2>
<section id="single-randomized-group" class="level3"><h3 class="anchored" data-anchor-id="single-randomized-group">Single, randomized group</h3>
<p>We will first start with looking at the most simple experimental design: a single, randomized group. Once we are comfortable with this, then we will return to a two group comparision in classic differential gene expression.</p>
<p>When we run experiments, we are usually analyzing a subset of the population, called a sample. Let’s define them clearly:</p>
<p><strong>Population:</strong> The entire collection of individual units that a researcher is interested to study.</p>
<p><strong>Sample:</strong> A smaller collection of individual units that the researcher has selected to study.</p>
<p>When we carry out our experiment, we perform some form of measurement on our samples, such as an experimental assay. Then, it is common to look at the mean and variance of our measurements to summarize our samples. This <strong>sample mean</strong> and <strong>sample variance</strong> are our best guess at the <strong>population mean</strong> and <strong>population variance</strong>.</p>
<p>There is a crucial relationship between samples, populations, and sample size: As the sample size gets larger, then the sample mean approaches the population mean and the sample variance approaches the population variance. This is called the <strong>Law of Large Numbers</strong>.</p>
<p>Let’s take at the following experiment as a demo: The goal of this <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96870">study</a> was to determine the effect of an upper-respiratory infection on changes in RNA transcription occurring in the cerebellum on mouse models. Control mice were inoculated with saline at Day 0 and measured on the same day. Case mice were inoculated with Influenza A and measured at day 8 post-infection.</p>
<p>Let’s take a look at RNA-seq measurement of the gene <code>Asl</code> for the control samples.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySummarizedExperiment">tidySummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSE96870</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">GSE96870_Case_Control</span> <span class="op">&lt;-</span> <span class="va">GSE96870</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">time</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Day0"</span>, <span class="st">"Day8"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_plot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">GSE96870_Case_Control</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.feature</span> <span class="op">==</span> <span class="st">"Asl"</span> <span class="op">&amp;</span> <span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Label</span>, y <span class="op">=</span> <span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">expression_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="resources/images/figure/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We have 7 samples here, and we would compute the mean to get the sample mean for the counts, approximating the population mean.</p>
<p>By these properties, it is obvious that the more samples, the better, one might say. But your choice of samples also has an impact on the population you are approximating!</p>
<p>Let’s consider the following thought experiment:</p>
<p>Would you rather: Sequence 7 replicates on the same cell line, or sequence one replicate each on 7 different cell lines?</p>
<p>What would the sample mean and sample variance of each experiment tell you?</p>
<p>In light of this example, let’s define two types of replicates:</p>
<ul>
<li><p><strong>Technical replicates:</strong> use the same biological sample to repeat the technical or experimental steps in order to accurately measure experimental variation.</p></li>
<li><p><strong>Biological replicates</strong> use different biological samples of the same condition to measure the biological variation between samples.</p></li>
</ul></section><section id="two-group-comparisons" class="level3"><h3 class="anchored" data-anchor-id="two-group-comparisons">Two-group comparisons</h3>
<p>In classical differntial gene expression, you are often conducting a two-group comparison between case and control. Our case samples arise from the case population, and our control samples arise from the control population. When we look at the measurements of a gene, we look at the <em>difference</em> of sample mean betewen case and controls, which is on a log-transformed scale called <strong>log-fold change</strong>. As the sample size gets larger, our gene’s sample log-fold change gets closer to the population log-fold change. Similarly, the sample variance for from case and control groups for this gene gets closer to their respective population variances.</p>
<p>In the ideal experimental set-up, you want all of your samples to be randomly sampled and differ only by the case and control conditions. However, that rarely happens in real life. There are other variables lurking, also known as <strong>confounding variables,</strong> that may also have an impact on the outcome you want to measure.</p>
<ul>
<li><p>Any relevant metadata, such as age, sex, subtype</p></li>
<li><p>Batch effects: the date of the experiment conducted, the person who performed the experiment, the reagents used, etc.</p></li>
<li><p>The sequencing technology</p></li>
<li><p>The amount of sequencing the sample received (more on this later in Normalization)</p></li>
<li><p>What else?</p></li>
</ul>
<p>This means you need to think carefully what the sample log-fold change are telling you: does the sample log-fold change relate more to your experimental condition, or something else, such as batch effects? A variable can <strong>confound</strong> your experiment when you can’t tell whether this variable or your experimental design is affecting the outcome.</p>
<p>Example: We know that sex has large effects on gene expression, and if all of our control mice were female and all of the treatment mice were male, then our treatment effect would be confounded by the variable sex. We could not differentiate the effect of treatment from the effect of sex.</p>
<p>Example: We know that batch effect has large effects on gene expression, and if all of our control samples were processed by an experienced researcher on old reagents, and our treatment samples were processed by an trainee researcher on new reagents, our treatment effect would be confounded by batch effects.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="https://web.stanford.edu/class/bios221/book/13-chap.html"><img src="https://web.stanford.edu/class/bios221/book/13-chap_files/figure-html/fig-confounding-1-1.png" class="img-fluid figure-img" width="350" alt="Source: Modern Statistics for Modern Biology, Chapter 13"></a></p>
<figcaption>Source: Modern Statistics for Modern Biology, Chapter 13</figcaption></figure>
</div>
</section></section><section id="dealing-with-confounding" class="level2"><h2 class="anchored" data-anchor-id="dealing-with-confounding">Dealing with confounding</h2>
<p>Solutions:</p>
<ul>
<li><p>Ensure that the samples in all experimental conditions have the same values for the confounding variable - ie. all of the same sex, all of the same batch.</p></li>
<li>
<p>If not possible, make sure that all experimental conditions have a similar amount of variability of the confounding variable - ie. the both case and control groups have a balance of males and females, and various batches.</p>
<ul>
<li><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="https://web.stanford.edu/class/bios221/book/13-chap_files/figure-html/fig-blockbox-1.png" class="img-fluid figure-img" width="350"></p>
<figcaption>Source: Modern Statistics for Modern Biology, Chapter 13</figcaption></figure>
</div></li>
<li>Save this information; we can eliminate these variability in the modeling process.</li>
</ul>
</li>
</ul>
<p>Let’s look at our experimental design. Have our potential confounders been accounted for?</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">GSE96870_Case_Control</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus 8 weeks Female
GSM2545342  CNS_RNA-seq_1C    GSM2545342 Mus musculus 8 weeks Female
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus 8 weeks Female
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus 8 weeks Female
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus 8 weeks   Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus 8 weeks   Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse          Label       Group
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9  Female_Day0_9 Female_Day0
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10 Female_Day0_10 Female_Day0
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8  Female_Day0_8 Female_Day0
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4  Female_Day0_4 Female_Day0
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11   Male_Day0_11   Male_Day0
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7    Male_Day0_7   Male_Day0
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2    Male_Day0_2   Male_Day0
GSM2545336  InfluenzaA C57BL/6 Day8 Cerebellum    14 Female_Day8_14 Female_Day8
GSM2545342  InfluenzaA C57BL/6 Day8 Cerebellum     5  Female_Day8_5 Female_Day8
GSM2545351  InfluenzaA C57BL/6 Day8 Cerebellum    16 Female_Day8_16 Female_Day8
GSM2545380  InfluenzaA C57BL/6 Day8 Cerebellum    19 Female_Day8_19 Female_Day8
GSM2545341  InfluenzaA C57BL/6 Day8 Cerebellum     6    Male_Day8_6   Male_Day8
GSM2545346  InfluenzaA C57BL/6 Day8 Cerebellum    23   Male_Day8_23   Male_Day8
GSM2545347  InfluenzaA C57BL/6 Day8 Cerebellum    24   Male_Day8_24   Male_Day8</code></pre>
</div>
</div>
</section><section id="specific-confounding-issues-in-rna-seq-experiments" class="level2"><h2 class="anchored" data-anchor-id="specific-confounding-issues-in-rna-seq-experiments">Specific confounding issues in RNA-seq experiments</h2>
<p>Besides looking at our metadata, there are two confounding variables that happen in nearly every RNA-seq experiment that people have spent a great deal of time developing tools to reduce confounding.</p>
<section id="library-size" class="level3"><h3 class="anchored" data-anchor-id="library-size">Library Size</h3>
<p>Each sample will receive different number of reads for sequencing, and that may confound the our desired measurement of log-fold change. The total number of reads a sample received for sequencing is called the <strong>library size</strong>.</p>
<p>The following plot is the plot of the expression counts for the <code>Asl</code> gene across all samples in our dataset grouped by <code>time</code>. There appears to be a difference in expression across the samples by timepoint, but is that due to the experimental condition (Day 0 is control, Day 8 is case), or due to the number of sequencing reads each sample received?</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_plot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">GSE96870_Case_Control</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.feature</span> <span class="op">==</span> <span class="st">"Asl"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Label</span>, y <span class="op">=</span> <span class="va">counts</span>, color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">expression_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="resources/images/figure/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The issue is that the library sizes of each sample is different. If we compare the counts of each library, we’ll see they vary.</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">expression_plot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">GSE96870_Case_Control</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.feature</span> <span class="op">==</span> <span class="st">"Asl"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Label</span>, y <span class="op">=</span> <span class="va">counts</span>, color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Asl Expression count"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">count_plot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">GSE96870_Case_Control</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Label</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>total_counts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Label</span>, y<span class="op">=</span><span class="va">total_counts</span>, fill<span class="op">=</span><span class="va">time</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Library Size for each Sample"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>tidySummarizedExperiment says: A data frame is returned for independent data analysis.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'Label'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_plot</span><span class="op">/</span><span class="va">count_plot</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="resources/images/figure/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You will notice that the counts for <code>Asl</code> across the samples vary by Library size (if you can’t see it, look at the patterns of ups and downs). For us to truly compare the expression differences between our timepoints in our samples, we need to remove the effect of <em>library size</em> before we make the comparison between case and control.</p>
<p>The easiest way to remove this confounder is to divide gene count of each sample by its respective total library size. However, there’s a second confounder that needs to be simultaneously addressed…</p>
</section><section id="rna-composition" class="level3"><h3 class="anchored" data-anchor-id="rna-composition">RNA Composition</h3>
<p>The second source of confounding has to do with the relative measurement of RNA-seq. Consider a single sample you just perforemd RNA-seq on. The assay provides a measure of the relative abundance of each gene in each RNA sample, but does not provide any measure of the total RNA output - it is a relative measurement. That is, the proportion of reads attributed to a given gene in a library depends on the expression properties of the whole sample rather than just the expression level of that gene.</p>
<p>Then, if a small proportion of highly expressed genes consume a substantial proportion of the total library size for a particular sample, this will cause the remaining genes to be under-sampled for that sample, confounding our analysis analysis.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="https://hbctraining.github.io/Intro-to-DGE/img/normalization_methods_composition.png" class="img-fluid figure-img" width="450"></p>
<figcaption>Source: Harvard Bioinformatics Training. Imagine the sequencing depths are similar between Sample A and Sample B, and every gene except for gene differentially expressed presents similar expression level between samples. The counts in Sample B would be greatly skewed by the differentially expressed gene, which takes up most of the counts. Other genes for Sample B would therefore appear to be less expressed than those same genes in Sample A.</figcaption></figure>
</div>
<p>The way we deal with this type of confounding is to make the assumption that <em>most</em> genes do not exhibit differential expression. DESeq2 divides gene count of each sample by its “size factor”, which is related to the sample’s library size for non-differentially expressed genes. If you want to see how it is calculated, this is a <a href="https://hbctraining.github.io/Intro-to-DGE/lessons/02_DGE_count_normalization.html">great tutorial</a>.</p>
</section></section><section id="putting-this-together-in-practice" class="level2"><h2 class="anchored" data-anchor-id="putting-this-together-in-practice">Putting this together in practice</h2>
<p>First, let’s do a little bit of clean up. Take a look at the total number of reads each gene received, and remove genes that just don’t have many reads across samples to be analyzed. Here’s the first five genes.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">GSE96870_Case_Control</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Xkr4 LOC105243853 LOC105242387 LOC105242467          Rp1 
       27023           12         2396           52           29 </code></pre>
</div>
</div>
<p>Let’s look at its distribution:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">GSE96870_Case_Control</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
      0       1      39   12144    5198 2042743 </code></pre>
</div>
</div>
<p>We started with 42k genes:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">GSE96870_Case_Control</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 41786    14</code></pre>
</div>
</div>
<p>Let’s remove any genes from the analysis if they received less than 5 reads total across samples.</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870_filtered</span> <span class="op">&lt;-</span> <span class="va">GSE96870_Case_Control</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">GSE96870_Case_Control</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">GSE96870_filtered</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 26818    14</code></pre>
</div>
</div>
<p>Which reduced us to 27k genes.</p>
<p>Now, let’s create a <code>DESeq</code> object from our <code>SummarizedExperiment</code> object and our experimental design. We need to provide an experimental design formula. The <strong>design formula</strong> should have all of the factors in your metadata that account for major sources of variation in your data.</p>
<p>Let’s look at our metadata to remind us all the variables in there:</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">GSE96870_filtered</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus 8 weeks Female
GSM2545342  CNS_RNA-seq_1C    GSM2545342 Mus musculus 8 weeks Female
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus 8 weeks Female
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus 8 weeks Female
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus 8 weeks   Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus 8 weeks   Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse          Label       Group
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9  Female_Day0_9 Female_Day0
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10 Female_Day0_10 Female_Day0
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8  Female_Day0_8 Female_Day0
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4  Female_Day0_4 Female_Day0
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11   Male_Day0_11   Male_Day0
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7    Male_Day0_7   Male_Day0
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2    Male_Day0_2   Male_Day0
GSM2545336  InfluenzaA C57BL/6 Day8 Cerebellum    14 Female_Day8_14 Female_Day8
GSM2545342  InfluenzaA C57BL/6 Day8 Cerebellum     5  Female_Day8_5 Female_Day8
GSM2545351  InfluenzaA C57BL/6 Day8 Cerebellum    16 Female_Day8_16 Female_Day8
GSM2545380  InfluenzaA C57BL/6 Day8 Cerebellum    19 Female_Day8_19 Female_Day8
GSM2545341  InfluenzaA C57BL/6 Day8 Cerebellum     6    Male_Day8_6   Male_Day8
GSM2545346  InfluenzaA C57BL/6 Day8 Cerebellum    23   Male_Day8_23   Male_Day8
GSM2545347  InfluenzaA C57BL/6 Day8 Cerebellum    24   Male_Day8_24   Male_Day8</code></pre>
</div>
</div>
<p>The variables that may contribute to variation to our data are: <code>sex</code>, <code>time</code>. There are variables such as <code>infection</code>, <code>Label</code>, <code>Group</code> that are encoded in <code>Label</code> and <code>Group</code> so we don’t need to include that information. The <code>mouse</code> variable is the unique identifier for each sample so that isn’t needed either.</p>
<p>Let’s create our <code>DESeq</code> object using <code>SummarizedExperiment</code> object and our design formula.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870_deseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSet</a></span><span class="op">(</span><span class="va">GSE96870_filtered</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in DESeqDataSet(GSE96870_filtered, design = ~sex + time): some
variables in design formula are characters, converting to factors</code></pre>
</div>
</div>
<p>Let’s calculate our “size factor” to deal with the library size and RNA composition confounders via the <code>estimateSizeFactors</code> function. Afterwards, you can see the size factor on the rightmost column of our colData.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GSE96870_normalized</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">GSE96870_deseq</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">GSE96870_normalized</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus 8 weeks Female
GSM2545342  CNS_RNA-seq_1C    GSM2545342 Mus musculus 8 weeks Female
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus 8 weeks Female
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus 8 weeks Female
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus 8 weeks   Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus 8 weeks   Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse          Label       Group
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9  Female_Day0_9 Female_Day0
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10 Female_Day0_10 Female_Day0
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8  Female_Day0_8 Female_Day0
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4  Female_Day0_4 Female_Day0
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11   Male_Day0_11   Male_Day0
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7    Male_Day0_7   Male_Day0
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2    Male_Day0_2   Male_Day0
GSM2545336  InfluenzaA C57BL/6 Day8 Cerebellum    14 Female_Day8_14 Female_Day8
GSM2545342  InfluenzaA C57BL/6 Day8 Cerebellum     5  Female_Day8_5 Female_Day8
GSM2545351  InfluenzaA C57BL/6 Day8 Cerebellum    16 Female_Day8_16 Female_Day8
GSM2545380  InfluenzaA C57BL/6 Day8 Cerebellum    19 Female_Day8_19 Female_Day8
GSM2545341  InfluenzaA C57BL/6 Day8 Cerebellum     6    Male_Day8_6   Male_Day8
GSM2545346  InfluenzaA C57BL/6 Day8 Cerebellum    23   Male_Day8_23   Male_Day8
GSM2545347  InfluenzaA C57BL/6 Day8 Cerebellum    24   Male_Day8_24   Male_Day8
           sizeFactor
GSM2545337  0.9133349
GSM2545338  0.8598908
GSM2545348  1.0456296
GSM2545353  1.1116468
GSM2545343  1.1217156
GSM2545349  1.0209117
GSM2545354  0.9843337
GSM2545336  1.1394808
GSM2545342  0.8802526
GSM2545351  1.0487852
GSM2545380  1.1395748
GSM2545341  0.8886238
GSM2545346  0.9389596
GSM2545347  1.0149250</code></pre>
</div>
</div>
<p>If we divide <code>counts</code> by <code>sizeFactor</code>, then we get our <code>scaled</code> counts.</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_plot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">GSE96870_Case_Control</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.feature</span> <span class="op">==</span> <span class="st">"Asl"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Label</span>, y <span class="op">=</span> <span class="va">counts</span>, color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">1100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Asl Expression count"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">expression_normalized</span> <span class="op">&lt;-</span> <span class="va">GSE96870_normalized</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.feature</span> <span class="op">==</span> <span class="st">"Asl"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>scaled <span class="op">=</span> <span class="va">counts</span> <span class="op">/</span> <span class="va">sizeFactor</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Label</span>, y <span class="op">=</span> <span class="va">scaled</span>, color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">1100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Asl Expression count normalized"</span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">expression_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="va">expression_normalized</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 3 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 2 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="resources/images/figure/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As you can see, some, but not all of the intra-timepoint variation was removed, especially within the <code>Day0</code> timepoint.</p>
<p>Whew, that’s all for experimental designs for now!</p>
<p>Here’s a nice summary of what we covered to ensure good experimental designs: “Generally speaking, a well-designed experiment is one that [has enough samples] and one in which technical artifacts and biological features that may systematically affect measurements are balanced, randomized or controlled in some other way in order to minimize opportunities for multiple explanations for the effect(s) under study.” (<a href="https://web.stanford.edu/class/bios221/book/16-chap.html#ref-Bacher:GB:2016">Bacher and Kendziorski 2016</a>)</p>
</section><section id="appendix-how-do-i-know-how-many-samples-to-use" class="level2"><h2 class="anchored" data-anchor-id="appendix-how-do-i-know-how-many-samples-to-use">Appendix: How do I know how many samples to use?</h2>
<p>Define “power”:</p>
<p>Reference to some packages for power calculations, such as <a href="https://bioconductor.org/packages/release/bioc/html/PROPER.html">PROPER</a>. Takes consideration of:</p>
<ul>
<li><p>Number of samples</p></li>
<li><p>Number of genes measured</p></li>
<li><p>Number of genes significant</p></li>
<li><p>Effect size of significant genes</p></li>
<li><p>p-value cutoff in multiple testing</p></li>
<li><p>Default datasets to simulate</p></li>
<li>
<p>Definition of power in multiple testing</p>
<ul>
<li>Probability we detect at least X% of genes that are truly different (at least some FC)</li>
</ul>
</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">PROPER</span><span class="op">)</span></span>
<span><span class="va">sim.opts.Cheung</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/PROPER/man/RNAseq.SimOptions.2grp.html">RNAseq.SimOptions.2grp</a></span><span class="op">(</span>ngenes <span class="op">=</span> <span class="fl">20000</span>, p.DE<span class="op">=</span><span class="fl">0.05</span>, lOD<span class="op">=</span><span class="st">"cheung"</span>, lBaselineExpr<span class="op">=</span><span class="st">"cheung"</span><span class="op">)</span></span>
<span><span class="va">simres</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/PROPER/man/runSims.html">runSims</a></span><span class="op">(</span>Nreps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">7</span>, <span class="fl">10</span><span class="op">)</span>, sim.opts<span class="op">=</span><span class="va">sim.opts.Cheung</span>, DEmethod<span class="op">=</span><span class="st">"DESeq2"</span>, nsims<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Simulation number 1 
Simulation number 2 
Simulation number 3 
Simulation number 4 
Simulation number 5 </code></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">powers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/PROPER/man/comparePower.html">comparePower</a></span><span class="op">(</span><span class="va">simres</span>, alpha.type<span class="op">=</span><span class="st">"fdr"</span>, alpha.nominal<span class="op">=</span><span class="fl">0.1</span>, stratify.by<span class="op">=</span><span class="st">"expr"</span>, delta<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/PROPER/man/summaryPower.html">summaryPower</a></span><span class="op">(</span><span class="va">powers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     SS1 SS2 Nominal FDR Actual FDR Marginal power Avg # of TD Avg # of FD  FDC
[1,]   3   3         0.1       0.36           0.39          60          35 0.57
[2,]   5   5         0.1       0.25           0.49          78          27 0.34
[3,]   7   7         0.1       0.20           0.55          89          23 0.26
[4,]  10  10         0.1       0.17           0.60          99          22 0.22</code></pre>
</div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/PROPER/man/plots.html">plotPower</a></span><span class="op">(</span><span class="va">powers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="resources/images/figure/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This suggests that you get more power out of more samples, rather than sequencing.</p>
<p>The most rigorous power benchmarks are to use controlled labatory experiments, with many samples and sequencing, and then downsample.</p>
</section><section id="appendix-dimension-reduction-to-explore-your-data" class="level2"><h2 class="anchored" data-anchor-id="appendix-dimension-reduction-to-explore-your-data">Appendix: Dimension Reduction to explore your data</h2>
</section><section id="appendix-other-confounding-factors-to-consider" class="level2"><h2 class="anchored" data-anchor-id="appendix-other-confounding-factors-to-consider">Appendix: Other confounding factors to consider</h2>
<ul>
<li><p>Unobserved confounders</p></li>
<li><p>Are the demarcation of groups I’m comparing is independent of my molecular measurements? (double dipping): https://stat.uw.edu/seminars/double-dipping-problems-and-solutions-application-single-cell-rna-sequencing-data, https://anna-neufeld.github.io/datathin/</p></li>
</ul></section></main><!-- /main column --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>